tests/unit/lms/validation/_lti_launch_params_test.py

```diff
@@ -119,7 +120,101 @@ def set_url_(url):
 
     @pytest.fixture
     def schema(self, pyramid_request):
-        return LaunchParamsURLConfiguredSchema(pyramid_request)
+        return URLConfiguredBasicLTILaunchSchema(pyramid_request)
+
+
+class TestContentItemSelectionLTILaunchSchema:
+    def test_it(self, schema):
+        parsed_params = schema.parse()
+
+        assert parsed_params == {
+            "context_id": "test_context_id",
+            "context_title": "test_context_title",
+            "lti_message_type": "ContentItemSelectionRequest",
+            "lti_version": "LTI-1p0",
+            "oauth_consumer_key": "test_oauth_consumer_key",
+            "tool_consumer_instance_guid": "test_tool_consumer_instance_guid",
+            "user_id": "test_user_id",
+            "custom_canvas_api_domain": "test_custom_canvas_api_domain",
+            "custom_canvas_course_id": "test_custom_canvas_course_id",
+            "launch_presentation_return_url": "test_launch_presentation_return_url",
+            "lis_person_name_full": "test_lis_person_name_full",
+            "lis_person_name_family": "test_lis_person_name_family",
+            "lis_person_name_given": "test_lis_person_name_given",
+            "tool_consumer_info_product_family_code": "test_tool_consumer_info_product_family_code",
+        }
+
+    @pytest.mark.parametrize(
+        "missing_param",
+        [
+            "context_id",
+            "context_title",
+            "lti_message_type",
+            "lti_version",
+            "oauth_consumer_key",
+            "tool_consumer_instance_guid",
+            "user_id",
+        ],
+    )
+    def test_required_params(self, schema, pyramid_request, missing_param):
+        del pyramid_request.params[missing_param]
+
+        with pytest.raises(ValidationError) as exc_info:
+            schema.parse()
+
+        assert exc_info.value.messages == dict(
+            ((missing_param, ["Missing data for required field."]),)
+        )
+
+    @pytest.mark.parametrize(
+        "invalid_params,expected_error_messages",
+        [
+            (
+                {"lti_version": "invalid version",},
+                {"lti_version": ["Must be one of: LTI-1p0."]},
```

Comment:
Sadly this probably should be pulled apart to put the `_CommonLTILaunchSchema` elsewhere, but the test code would probably be about the same size if both were inlined as the test is nice and short.

Reply:
Yeah it's a tradeoff. You get to remove some test duplication, but at the cost of introducing some test cleverness. None of the solutions I've seen (to sharing a set of tests among multiple classes or methods-under-test) are great